proposals/NNNN-asyncsequence.md

```diff
@@ -0,0 +1,386 @@
+
+```swift
+var allLines: [String] = []
+do {
+  for try await line in myFile.lines() {
+    allLines.append(line)
+  }
+} catch {
+  allLines = []
+}
+```
+
+There's nothing wrong with the above code, and it must be possible for a developer to write it. However, it does seem like a lot of boilerplate for what might be a common operation. One way to solve this would be to add more functions to `URL`:
+
+```swift
+extension URL {
+  struct Lines : AsyncSequence { }
+
+  func lines() -> Lines
+  func firstLine() throws async -> String?
+  func collectLines() throws async -> [String]
+}
+```
+
+It doesn't take much imagination to think of other places where we may want to do similar operations, though. Therefore, we believe the best place to put these functions is instead as an extension on `AsyncSequence` itself, specified generically -- just like `Sequence`.
+
+## Proposed solution
+
+The standard library will define the following protocols:
+
+```swift
+public protocol AsyncSequence {
+  associatedtype AsyncIterator: AsyncIteratorProtocol where AsyncIterator.Element == Element
+  associatedtype Element
+  func makeAsyncIterator() -> AsyncIterator
+}
+
+public protocol AsyncIteratorProtocol {
+  associatedtype Element
+  mutating func next() async throws -> Element?
+  __consuming mutating func cancel()
+}
+```
+
+The compiler will generate code to allow use of a `for in` loop on any type which conforms with `AsyncSequence`. The standard library will also extend the protocol to provide familiar generic algorithms. Here is an example which does not actually call an `async` function within its `next`, but shows the basic shape:
+
+```swift
+struct Counter : AsyncSequence {
+  let howHigh: Int
+
+  struct AsyncIterator : AsyncIteratorProtocol {
+    let howHigh: Int
+    var current = 1
+    mutating func next() async -> Int? {
+      guard current <= howHigh else {
+        return nil
+      }
+
+      let result = current
+      current += 1
+      return result
+    }
+
+    mutating func cancel() {
+      current = howHigh + 1 // Make sure we do not emit another value
+    }
+  }
+
+  func makeAsyncIterator() -> AsyncIterator {
+    return AsyncIterator(howHigh: howHigh)
+  }
+}
+```
+
+At the call site, using `Counter` would look like this:
+
+```swift
+for await i in Counter(howHigh: 3) {
+  print(i)
+}
+
+/* 
+Prints the following, and finishes the loop:
+1
+2
+3
+*/
+
+
+for await i in Counter(howHigh: 3) {
+  print(i)
+  if i == 2 { break }
+}
+/*
+Prints the following, then calls cancel before breaking out of the loop:
+1
+2
+*/
+```
+
+Any other exit (e.g., `return` or `throw`) from the `for` loop will also call `cancel` first.
+
+## Detailed design
+
+Returning to our earlier example:
+
+```swift
+for try await line in myFile.lines() {
+  // Do something with each line
+}
+```
+
+The compiler will emit the equivalent of the following code:
+
+```swift
+var it = myFile.lines().makeAsyncIterator()
+while let value = try await it.next() {
```

Comment:
It seems too late but... Isn’t this line expected to be:

```swift
while let line = try await it.next() {
```

Reply:
Thank you -- I'll fix this in an update.